### ISSUE TITLE
[JENKINS-71492] Postprocessing hangs for minutes between "Successfully processed file" and "Post processing issues"

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/warnings-ng-plugin#3119]
---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=martinbauer">martinbauer</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492" target="_blank">Postprocessing hangs for minutes between "Successfully processed file" and "Post processing issues"</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drulli">drulli</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: warnings-ng-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2024-01-11T08:26:26+00:00
<li><b>votes</b>: 1
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-09
</ul></i>
</details>
<ul><li><i>environment</i>: <code>Windows Server 2019, Java 17, Jenkins LTS 2.401.1</code></li></ul>
<details><summary><i>1 attachment</i></summary>

- [ngw-3x.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/60628/ngw-3x.jpg)
  > ![ngw-3x.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/60628/ngw-3x.jpg)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [PR #​1656](https://github.com/jenkinsci/warnings-ng-plugin/pull/1656)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>martinbauer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=438822&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438822">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think I have found the cause.
Among other things, we used pnpm as a package manager.
This uses symlinks to save disk space with a central package store.
<a href="https://pnpm.io/symlinked-node-modules-structure" class="external-link" target="_blank" rel="nofollow noopener">https://pnpm.io/symlinked-node-modules-structure</a>
On Windows, junctions are used for this.
<a href="https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions" class="external-link" target="_blank" rel="nofollow noopener">https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions</a>
When I delete the node-modules folder which contains all the juctions, there are only a few seconds between the log lines.</p>

<p>I just don't understand why the plugin interacts with the folder at all. No issue from a report references a file from that folder.
And the folder is not inside the "Report File Pattern".
I also have disabled the setting for symbolic links:
<em>Traversing of symbolic links: disabled</em></p>
</pre>
</details>

</ul>
</details>

<p>I think I have found the cause.<br/>
Among other things, we used pnpm as a package manager.<br/>
This uses symlinks to save disk space with a central package store.<br/>
<a href="https://pnpm.io/symlinked-node-modules-structure" class="external-link" target="_blank" rel="nofollow noopener">https://pnpm.io/symlinked-node-modules-structure</a><br/>
On Windows, junctions are used for this.<br/>
<a href="https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions" class="external-link" target="_blank" rel="nofollow noopener">https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions</a><br/>
When I delete the node-modules folder which contains all the juctions, there are only a few seconds between the log lines.</p>

<p>I just don't understand why the plugin interacts with the folder at all. No issue from a report references a file from that folder.<br/>
And the folder is not inside the "Report File Pattern".<br/>
I also have disabled the setting for symbolic links:<br/>
<em>Traversing of symbolic links: disabled</em></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=441563&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-441563">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It seems a problem with the post processing step. Currently the log messages of that step are printed at the same time so it is hard to see which of these steps cause the problems. </p>

<p>A first guess would be the module detector, that scans the workspace for module names. In case of a node structure this might take too long. </p>
</pre>
</details>

</ul>
</details>

<p>It seems a problem with the post processing step. Currently the log messages of that step are printed at the same time so it is hard to see which of these steps cause the problems. </p>

<p>A first guess would be the module detector, that scans the workspace for module names. In case of a node structure this might take too long. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=441591&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-441591">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>One helpful thing you can do is to create a thread dump of agent and controller while the processes are hanging in this 5 minute gap.</p>
</pre>
</details>

</ul>
</details>

<p>One helpful thing you can do is to create a thread dump of agent and controller while the processes are hanging in this 5 minute gap.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gustawsmolarczyk:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=441606&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-441606">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello.</p>

<p>We have encountered a similar problem with a different use case. What I have analyzed might also help the original issue.</p>

<p>We have a C++ application and we run scanForIssues() on a gcc- or clang-produced log files. Scanning hangs for 5-15 minutes between the above-mentioned two lines. When I looked at the thread dump (not the job thread dump but the whole Jenkins controller one) I have seen a thread:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">	at java.base@11.0.14.1/java.io.UnixFileSystem.list(Native Method)
	at java.base@11.0.14.1/java.io.File.normalizedList(Unknown Source)
	at java.base@11.0.14.1/java.io.File.list(Unknown Source)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1252)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1198)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1160)
	at org.apache.tools.ant.DirectoryScanner.checkIncludePatterns(DirectoryScanner.java:954)
	at org.apache.tools.ant.DirectoryScanner.scan(DirectoryScanner.java:912)
	-  locked org.apache.tools.ant.DirectoryScanner@36457bf1
	at org.apache.tools.ant.types.AbstractFileSet.getDirectoryScanner(AbstractFileSet.java:528)
	at io.jenkins.plugins.analysis.core.util.FileFinder.find(FileFinder.java:110)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$DefaultFileSystem.find(IssuesScanner.java:363)
	at edu.hm.hafner.analysis.ModuleDetector.find(ModuleDetector.java:114)
	at edu.hm.hafner.analysis.ModuleDetector.createFilesToModuleMapping(ModuleDetector.java:73)
	at edu.hm.hafner.analysis.ModuleDetector.&lt;init&gt;(ModuleDetector.java:57)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.resolveModuleNames(IssuesScanner.java:320)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.invoke(IssuesScanner.java:268)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.invoke(IssuesScanner.java:239) </pre>
</div></div>
<p> </p>

<p>I have found that resolveModuleNames() creates a ModuleDetector here:</p>

<p><a href="https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java#L315" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java#L315</a></p>

<p>And then in the constructor it searches for a number of files in the whole workspace:</p>

<p><a href="https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L50" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L50</a>
<a href="https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L110" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L110</a></p>

<p>Our workspace contains none of the files being looked for (build.xml, etc), but the act of searching takes a long time itself.</p>

<p>The underlying workspace file system has high latency because it is hosted on an NFS - synchronous file system operations are thus really slow (on the other of milliseconds). The source tree is really large and traversing it takes quite a long time. The agent also runs on a separate host (more specifically, on a separate kubernetes pod) which might contribute to the slowness.</p>

<p>What would help us is the ability to opt-out of the workspace scanning. As mentioned, we do not have any of the files being looked for, so the whole exercise just wastes time. I don't think it would help the original issue though.</p>

<p>Another thing that could help is to make the search process asynchronous. While a single NFS operation (directory listing, stat, etc) has a relatively high latency, multiple such operations can be scheduled in parallel to hide this latency and improve the whole search process tremendously. Not sure how feasible it is as currently org.apache.tools.ant.types.FileSet is being used directly and it works synchronously.</p>
</pre>
</details>

</ul>
</details>

<p>Hello.</p>

<p>We have encountered a similar problem with a different use case. What I have analyzed might also help the original issue.</p>

<p>We have a C++ application and we run scanForIssues() on a gcc- or clang-produced log files. Scanning hangs for 5-15 minutes between the above-mentioned two lines. When I looked at the thread dump (not the job thread dump but the whole Jenkins controller one) I have seen a thread:</p>

<pre>
	at java.base@​11.0.14.1/java.io.UnixFileSystem.list(Native Method)
	at java.base@​11.0.14.1/java.io.File.normalizedList(Unknown Source)
	at java.base@​11.0.14.1/java.io.File.list(Unknown Source)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1252)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1282)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1198)
	at org.apache.tools.ant.DirectoryScanner.scandir(DirectoryScanner.java:1160)
	at org.apache.tools.ant.DirectoryScanner.checkIncludePatterns(DirectoryScanner.java:954)
	at org.apache.tools.ant.DirectoryScanner.scan(DirectoryScanner.java:912)
	-  locked org.apache.tools.ant.DirectoryScanner@​36457bf1
	at org.apache.tools.ant.types.AbstractFileSet.getDirectoryScanner(AbstractFileSet.java:528)
	at io.jenkins.plugins.analysis.core.util.FileFinder.find(FileFinder.java:110)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$DefaultFileSystem.find(IssuesScanner.java:363)
	at edu.hm.hafner.analysis.ModuleDetector.find(ModuleDetector.java:114)
	at edu.hm.hafner.analysis.ModuleDetector.createFilesToModuleMapping(ModuleDetector.java:73)
	at edu.hm.hafner.analysis.ModuleDetector.<init>(ModuleDetector.java:57)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.resolveModuleNames(IssuesScanner.java:320)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.invoke(IssuesScanner.java:268)
	at io.jenkins.plugins.analysis.core.steps.IssuesScanner$ReportPostProcessor.invoke(IssuesScanner.java:239) </pre>
<p> </p>

<p>I have found that resolveModuleNames() creates a ModuleDetector here:</p>

<p><a href="https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java#L315" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java#L315</a></p>

<p>And then in the constructor it searches for a number of files in the whole workspace:</p>

<p><a href="https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L50" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L50</a><br/>
<a href="https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L110" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/ModuleDetector.java#L110</a></p>

<p>Our workspace contains none of the files being looked for (build.xml, etc), but the act of searching takes a long time itself.</p>

<p>The underlying workspace file system has high latency because it is hosted on an NFS - synchronous file system operations are thus really slow (on the other of milliseconds). The source tree is really large and traversing it takes quite a long time. The agent also runs on a separate host (more specifically, on a separate kubernetes pod) which might contribute to the slowness.</p>

<p>What would help us is the ability to opt-out of the workspace scanning. As mentioned, we do not have any of the files being looked for, so the whole exercise just wastes time. I don't think it would help the original issue though.</p>

<p>Another thing that could help is to make the search process asynchronous. While a single NFS operation (directory listing, stat, etc) has a relatively high latency, multiple such operations can be scheduled in parallel to hide this latency and improve the whole search process tremendously. Not sure how feasible it is as currently org.apache.tools.ant.types.FileSet is being used directly and it works synchronously.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gustawsmolarczyk:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=444319&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-444319">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drulli" class="user-hover" rel="drulli">drulli</a> </p>

<p>Thank you for taking a look at this.</p>

<p> </p>

<p>I have tried the new `skipPostProcessing` parameter and indeed it resolves the delay. Unfortunately, it has a side effect of removing links to the source files from the warning list. Specifically, on the `File` column there were previously links that pointed to source view of a specific file and line, with the warning presented inline, but now the column has no links. I would assume that generating this output and the links is part of the post-processing.</p>

<p> </p>

<p>I would have preferred if the links and the annotated sources stayed and only searching for the module files (`build.xml`, etc) was skipped. I am not sure how feasible that is - do we need to scan in order to find each and every file referenced by the warnings? Or does it perform one FS lookup and read per file instead? The latter should be a lot faster for us, and the delay would scale with the number of warnings/unique files.</p>

<p> </p>

<p>While I was experimenting before, I already ran the `scanForIssues` step with an empty workspace so that the scan would not trigger. It also had a similar side-effect of removing links. So this new option just makes this work-around easier, but the end result is not what we want.</p>

<p> </p>

<p>Should I open another issue for this problem?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drulli" class="user-hover" rel="drulli">drulli</a> </p>

<p>Thank you for taking a look at this.</p>

<p> </p>

<p>I have tried the new `skipPostProcessing` parameter and indeed it resolves the delay. Unfortunately, it has a side effect of removing links to the source files from the warning list. Specifically, on the `File` column there were previously links that pointed to source view of a specific file and line, with the warning presented inline, but now the column has no links. I would assume that generating this output and the links is part of the post-processing.</p>

<p> </p>

<p>I would have preferred if the links and the annotated sources stayed and only searching for the module files (`build.xml`, etc) was skipped. I am not sure how feasible that is - do we need to scan in order to find each and every file referenced by the warnings? Or does it perform one FS lookup and read per file instead? The latter should be a lot faster for us, and the delay would scale with the number of warnings/unique files.</p>

<p> </p>

<p>While I was experimenting before, I already ran the `scanForIssues` step with an empty workspace so that the scan would not trigger. It also had a similar side-effect of removing links. So this new option just makes this work-around easier, but the end result is not what we want.</p>

<p> </p>

<p>Should I open another issue for this problem?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=444329&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-444329">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote>
<p>I would have preferred if the links and the annotated sources stayed and only searching for the module files (`build.xml`, etc) was skipped. I am not sure how feasible that is - do we need to scan in order to find each and every file referenced by the warnings? Or does it perform one FS lookup and read per file instead? The latter should be a lot faster for us, and the delay would scale with the number of warnings/unique files.</p></blockquote>

<p>This is a side effect as I was lazy and disabled the whole post processing. I can change that to skip only the package and module scanners. </p>
</pre>
</details>

</ul>
</details>

<blockquote>
<p>I would have preferred if the links and the annotated sources stayed and only searching for the module files (`build.xml`, etc) was skipped. I am not sure how feasible that is - do we need to scan in order to find each and every file referenced by the warnings? Or does it perform one FS lookup and read per file instead? The latter should be a lot faster for us, and the delay would scale with the number of warnings/unique files.</p></blockquote>

<p>This is a side effect as I was lazy and disabled the whole post processing. I can change that to skip only the package and module scanners. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=444335&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-444335">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I changed the behavior now. Can you please retry?</p>
</pre>
</details>

</ul>
</details>

<p>I changed the behavior now. Can you please retry?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gustawsmolarczyk:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-71492?focusedId=444336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-444336">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Works like a charm, thank you very much!</p>

<p> </p>

<p>Post-processing is now not skipped but takes less than 1 second (previously it was 1-2 minutes). The warnings list looks the same, so nothing was lost to gain this benefit.</p>
</pre>
</details>

</ul>
</details>

<p>Works like a charm, thank you very much!</p>

<p> </p>

<p>Post-processing is now not skipped but takes less than 1 second (previously it was 1-2 minutes). The warnings list looks the same, so nothing was lost to gain this benefit.</p>
